model/Image.php

```diff
@@ -328,6 +341,123 @@ public function generateSetSize(Image_Backend $backend, $width, $height) {
+		
+		return $this->getWidth() > $width || $this->getHeight() > $height
+			? $this->Fit($width,$height)
+			: $this;
+	}
+
+	/**
+	 * Proportionally scale down this image if it is wider than the specified width. 
+	 * Similar to ScaleWidth but without up-sampling. Use in templates with $ScaleMaxWidth.
+	 *
+	 * @uses Image::ScaleWidth()
+	 * @param integer $width The maximum width of the output image
+	 * @return Image
+	 */
+	public function ScaleMaxWidth($width) {
+		// Temporary $force_resample support for 3.x, to be removed in 4.0
+		if (Config::inst()->get('Image', 'force_resample') && $this->getWidth() <= $width) return $this->ScaleWidth($this->getWidth());
+		
+		return $this->getWidth() > $width
+			? $this->ScaleWidth($width)
+			: $this;
+	}
+
+	/**
+	 * Proportionally scale down this image if it is taller than the specified height. 
+	 * Similar to ScaleHeight but without up-sampling. Use in templates with $ScaleMaxHeight.
+	 *
+	 * @uses Image::ScaleHeight()
+	 * @param integer $height The maximum height of the output image
+	 * @return Image
+	 */
+	public function ScaleMaxHeight($height) {
+		// Temporary $force_resample support for 3.x, to be removed in 4.0
+		if (Config::inst()->get('Image', 'force_resample') && $this->getHeight() <= $height) return $this->ScaleHeight($this->getHeight());
+		
+		return $this->getHeight() > $height
+			? $this->ScaleHeight($height)
+			: $this;
+	}
+
+	/**
+	 * Crop this image to the aspect ratio defined by the specified width and height, 
+	 * then scale down the image to those dimensions if it exceeds them.
+	 * Similar to Fill but without up-sampling. Use in templates with $FillMax.
+	 *
+	 * @uses Image::Fill()
+	 * @param integer $width The relative (used to determine aspect ratio) and maximum width of the output image
+	 * @param integer $height The relative (used to determine aspect ratio) and maximum height of the output image
+	 * @return Image
+	 */
+	public function FillMax($width, $height) {
+		// Prevent divide by zero on missing/blank file
+		if(!$this->getWidth() || !$this->getHeight()) return null;
+		
+		// Temporary $force_resample support for 3.x, to be removed in 4.0
+		if (Config::inst()->get('Image', 'force_resample') && $this->isSize($width, $height)) return $this->Fill($width, $height);
+		
+		// Is the image already the correct size?
+		if ($this->isSize($width, $height)) return $this;
+		
+		// If not, make sure the image isn't upsampled
+		$imageRatio = $this->getWidth() / $this->getHeight();
+		$cropRatio = $width / $height;
+		// If cropping on the x axis compare heights
+		if ($cropRatio < $imageRatio && $this->getHeight() < $height) return $this->Fill($this->getHeight()*$cropRatio, $this->getHeight());
+		// Otherwise we're cropping on the y axis (or not cropping at all) so compare widths
+		if ($this->getWidth() < $width) return $this->Fill($this->getWidth(), $this->getWidth()/$cropRatio);
```

Comment:
Resurrecting our discussion from https://github.com/silverstripe/silverstripe-framework/pull/4229#discussion_r31980301

You said

> Ergh... Aargh... My brain is hurting too. Line 422 catches images that are going to be cropped on the x axis but are undersize and changes the resize arguments to suit. Line 424 catches images that are going to be cropped on the y axis AND images that aren't going to be cropped at all but are undersize. Checking the width only is sufficient at this point to cover both remaining cases because if the image isn't going to be cropped (the $imageRatio==$cropRatio scenario) it doesn't matter if we look at width or height, either one can tell us if the image is undersize. So... I'm 96% confident the code doesn't need changing :) but, sigh, maybe we need more unit tests.

Can you show me where `line 422` is in the rebased code so I can review?